src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1077,274 +897,288 @@ void FileCache::loadCacheInfoIntoMemory(std::lock_guard<std::mutex> & cache_lock
     }
 
-    CreateFileSegmentSettings create_settings(file_segment->getKind());
-
-    cell->file_segment = std::make_shared<FileSegment>(
-        offset, downloaded_size, key, this, FileSegment::State::DOWNLOADED, create_settings);
+    auto key_metadata = it->second;
+    auto key_guard = key_metadata->guard;
+    auto key_lock = key_guard->lock();
+
+    /// Removal of empty key from cache metadata is delayed.
+    /// Therefore here we need to check if the key which we currently took from cache metadata is a valid non-removed key.
+    /// Additional explanation:
+    ///   KeyMetadata::removed can be changed only under according KeyGuard::Lock. keyMetadata::removed is needed
+    ///   for delayed metadata cleanup, because usually we manage key metadata under key lock, e.g. delete the
+    ///   key metadata objects as well. So we could delete all key objects and the see that we deleted everything
+    ///   that we had for this key, so we consider that we need to remove this key from cache metadata. But in order
+    ///   to do that we need CacheMetadataGuard::Lock, which cannot be taken while holding KeyGuard::Lock.
+    ///   Therefore we need to retake locks in needed order to removed empty key from cache metadata. But it makes
+    ///   sense to do that with delay here.
+    if (key_metadata->removed)
+    {
+        /// If it is removed, then do perform delayed removal.
+        cleanup_keys_metadata_queue->remove(key);
+        metadata.erase(it);
+        /// Get metadata one more time.
+        /// Throw exception if KeyNotFoundPolicy::THROW, return nullptr if KeyNotFoundPolicy::RETURN_NULL.
+        if (key_not_found_policy == KeyNotFoundPolicy::THROW)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key `{}` in cache", key.toString());
+        else if (key_not_found_policy == KeyNotFoundPolicy::RETURN_NULL)
+            return nullptr;
+
+        it = metadata.emplace(key, std::make_shared<KeyMetadata>()).first;
+        return std::make_unique<LockedKey>(key, it->second, it->second->guard->lock(), cleanup_keys_metadata_queue, this);
+    }
 
-    assert(file_segment->reserved_size == downloaded_size);
+    return std::make_unique<LockedKey>(key, key_metadata, std::move(key_lock), cleanup_keys_metadata_queue, this);
 }
 
-bool FileCache::isLastFileSegmentHolder(
-    const Key & key, size_t offset,
-    std::lock_guard<std::mutex> & cache_lock, std::unique_lock<std::mutex> & /* segment_lock */)
+LockedKeyPtr FileCache::createLockedKey(const Key & key, KeyMetadataPtr key_metadata) const
 {
-    auto * cell = getCell(key, offset, cache_lock);
-
-    if (!cell)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "No cell found for key: {}, offset: {}", key.toString(), offset);
-
-    /// The caller of this method is last file segment holder if use count is 2 (the second pointer is cache itself)
-    return cell->file_segment.use_count() == 2;
+    auto key_lock = key_metadata->guard->lock();
```

Comment:
it cannot be done, because we must save shared pointer to a local variable, because below it is possible that we will remove this shared ptr from cache metadata (because of delayed removal), which means that the lock we currently took will result in segfault when it will try to unlock in destructor of unique lock.